libs/ml/include/ml/optimisation/adam_optimiser.hpp

```diff
@@ -0,0 +1,150 @@
+    : Optimiser<T, C>(graph, input_node_name, output_node_name, learning_rate)
+    ,
+
+    beta1_(beta1)
+    , beta2_(beta2)
+    , beta1_t_(beta1)
+    , beta2_t_(beta2)
+    , epsilon_(epsilon)
+  {
+    for (auto &train : this->graph_trainables_)
+    {
+      this->cache_.push_back(ArrayType(train->GetWeights().shape()));
+      this->momentum_.push_back(ArrayType(train->GetWeights().shape()));
+      this->mt_.push_back(ArrayType(train->GetWeights().shape()));
+      this->vt_.push_back(ArrayType(train->GetWeights().shape()));
+    }
+    ResetCache();
+  }
+
+private:
+  void ApplyGradients()
+  {
+    // Do operation with gradient
+    auto cached_weight_it = cache_.begin();
+    auto momentum_it      = momentum_.begin();
+    auto mt_it            = mt_.begin();
+    auto vt_it            = vt_.begin();
+
+    auto gradient_it  = this->gradients_.begin();
+    auto trainable_it = this->graph_trainables_.begin();
+
+    while (gradient_it != this->gradients_.end())
+    {
+      // cache[i] = (beta1_t_ * cache[i]) + ((one_ - beta1_t_) * (gradients[i]));
+      fetch::math::Multiply((*trainable_it)->Gradients(), (one_ - beta1_t_), *gradient_it);
+      fetch::math::Multiply(*cached_weight_it, beta1_t_, *cached_weight_it);
+      fetch::math::Add(*cached_weight_it, *gradient_it, *cached_weight_it);
+
+      // mt   = cache[i] / (one_ - beta1_t_);
+      fetch::math::Divide(*cached_weight_it, (one_ - beta1_t_), *mt_it);
+
+      // momentum[i] = (beta2_t_ * momentum[i]) + ((one_ - beta2_t_) * (gradients[i]^2));
+      fetch::math::Multiply((*trainable_it)->Gradients(), (*trainable_it)->Gradients(), *vt_it);
+      fetch::math::Multiply(*vt_it, (one_ - beta2_t_), *vt_it);
+      fetch::math::Multiply(*momentum_it, beta2_t_, *momentum_it);
+      fetch::math::Add(*momentum_it, *vt_it, *momentum_it);
+
+      // vt   = momentum[i] / (one_ - beta2_t_);
+      fetch::math::Divide(*momentum_it, (one_ - beta2_t_), *vt_it);
+
+      // gradients[i] = -this->learning_rate_ * mt / (sqrt(vt) + epsilon_);
+      fetch::math::Sqrt(*vt_it, *gradient_it);
+      fetch::math::Add(*gradient_it, epsilon_, *gradient_it);
+      fetch::math::Divide(*mt_it, *gradient_it, *gradient_it);
+      fetch::math::Multiply(*gradient_it, -this->learning_rate_, *gradient_it);
+
+      // Apply gradient weights[i]+=grad[i]
+      (*trainable_it)->ApplyGradient(*gradient_it);
+
+      ++cached_weight_it;
+      ++momentum_it;
+      ++mt_it;
+      ++vt_it;
+
+      ++gradient_it;
+      ++trainable_it;
+    }
+
+    // beta1_t=beta1^t and beta2_t=beta2^t, where t is number of epochs + 1
+    beta1_t_ *= beta1_;
+    beta2_t_ *= beta2_;
+  }
+
+  void ResetCache()
+  {
+    for (auto &val : this->cache_)
+    {
+      val.Fill(DataType{0});
+    }
+    for (auto &moment : this->momentum_)
+    {
+      moment.Fill(DataType{0});
+    }
+    beta1_t_ = beta1_;
+    beta2_t_ = beta2_;
+  }
+
+  std::vector<ArrayType> cache_;
+  std::vector<ArrayType> momentum_;
```

Comment:
in general it's nicer to have the member variables at the top too (so when you read a function you already know which are member variables); this is made easier when the function definitions are outside of the class since the class body becomes something like this:

class Name
{
  public:
    .. public variables
    .. public method declarations

  private:
    .. private variables
    .. private method declarations
}
.. public definitions
.. private definitions